libz-rs-sys/src/gz.rs

```diff
@@ -1296,6 +1303,340 @@ unsafe fn gz_decomp(state: &mut GzState) -> Result<(), ()> {
+        // Safety: we confirmed above that state is valid.
+        unsafe {
+            gz_error(
+                state,
+                Some((Z_DATA_ERROR, "requested length does not fit in int")),
+            )
+        };
+        return 0;
+    }
+
+    // Safety: We validated state above, and the caller is responsible for ensuring
+    // that buf points to at least len bytes of readable memory.
+    unsafe { gz_write(state, buf, len) }
+}
+
+// Internal implementation of `gzwrite`.
+//
+// # Returns
+// - The number of uncompress bytes written, on success.
+// - Or 0 in case of error.
+//
+// # Safety
+//
+/// - `state` must have been properly initialized, e.g. by [`gzopen_help`].
+/// - `buf` must point to at least `len` bytes of readable memory.
+/// - The caller must ensure that `len` fits in a `c_int`.
+unsafe fn gz_write(state: &mut GzState, mut buf: *const c_void, mut len: c_uint) -> c_int {
+    // If len is zero, avoid unnecessary operations.
+    if len == 0 {
+        return 0;
+    }
+
+    // Allocate memory if this is the first time through.
+    if state.input.is_null() && gz_init(state).is_err() {
+        return 0;
+    }
+
+    /* FIXME uncomment once seek support is implemented
+    if state.seek {
+        state.seek = 0;
+        if gz_zero(state, state.skip) == -1 {
+            return 0;
+        }
+    }
+     */
+
+    let put = len as c_int;
+
+    // For small len, copy to input buffer, otherwise compress directly.
+    if (len as usize) < state.in_size {
+        // Copy to input buffer, compress when full.
+        loop {
+            if state.stream.avail_in == 0 {
+                state.stream.next_in = state.input;
+            }
+            let have = unsafe {
+                state
+                    .stream
+                    .next_in
+                    .add(state.stream.avail_in as usize)
+                    .offset_from(state.input)
+            } as usize;
+            let copy = cmp::min(state.in_size.saturating_sub(have), len as usize);
+            // Safety: The caller is responsible for ensuring that buf points to at least len readable
+            // bytes, and copy is <= len.
+            unsafe { ptr::copy(buf, state.input.add(have).cast::<c_void>(), copy) };
+            state.stream.avail_in += copy as c_uint;
+            state.pos += copy as u64;
+            buf = unsafe { buf.add(copy) };
+            len -= copy as c_uint;
+            if len != 0 && gz_comp(state, Z_NO_FLUSH).is_err() {
+                return 0;
+            }
+            if len == 0 {
+                break;
+            }
+        }
+    } else {
+        // Consume any data left in the input buffer.
+        if state.stream.avail_in != 0 && gz_comp(state, Z_NO_FLUSH).is_err() {
+            return 0;
+        }
+
+        // Directly compress user buffer to file.
+        state.stream.next_in = buf.cast::<_>();
+        loop {
+            let n = cmp::min(len, c_uint::MAX);
```

Comment:
This seems kind of pointless? In the original code, `len` is a `size_t`, which I think can be 64-bit when `c_uint` is only 32-bit.

Reply:
I'll change the `len` parameter in this function to `usize` for consistency with the `size_t` in the original, which will make this check relevant.

Reply:
Actually when `len` is a `usize` this operation is actually meaningful.